### PR TITLE
Frontmatter as Page Context

### LINKED ISSUE #17: Frontmatter as PageContext
What do you think of adding the frontmatter automatically as `pageContext`? That way you could use those values in the GraphQL queries or just render them out in the Layout.

In `pages/index.mdx`
```js
---
title: "My post title"
---

Welcome to my blog.


export default ({ pageContext, children }) => (
  <div>
    <h1>{pageContext.frontmatter.title}</h1>
    {children}
  </div>
)
```

### PR DESCRIPTION
**Pull request by avigoldman**
Closes #17 

The only issue here is that `onPageCreate` only runs on the first cycle in dev mode so if the frontmatter changes while developing, the frontmatter in the `pageContext` doesn't update.

### CODE REVIEW

File: packages/gatsby-mdx/gatsby-node.js [hunk 1]
```diff
@@ -13,6 +16,38 @@ exports.onCreateNode = require("./on-create-node");
...
+    // grab the exported frontmatter
+    const { frontmatter } = extractExports(code);
+
+    deletePage(page);
+    createPage(
+      merge(
+        {
+          context: {
+            frontmatter: {
+              title: ``, // always include a title
```
Comment (ChristopherBiscardi) [inline packages/gatsby-mdx/gatsby-node.js]: why always include a title? Seems like we're assuming a blog post content type or something

File: examples/kitchen-sink/src/pages/basic-page-query.mdx [hunk 2]
```diff
@@ -1,5 +1,13 @@
+---
+my: frontmatter
+---
+
 import { graphql } from 'gatsby';
 
+export const frontmatter = {
+  other: 'frontmatter'
```
Comment (ChristopherBiscardi) [inline examples/kitchen-sink/src/pages/basic-page-query.mdx]: if there's going to be frontmatter here, I'd love for it to be used in the query and layout or something as well to solidify the use-case.

File: packages/gatsby-mdx/gatsby-node.js (hunk 1 above)
Comment (avigoldman) [reply packages/gatsby-mdx/gatsby-node.js]: I was going for consistency with the Mdx node. I don't have a particular use case in mind for it, so I'll take it out.

**Comment by avigoldman** [comment]
@ChristopherBiscardi ready for a re-review.

File: packages/gatsby-mdx/on-create-node.js [hunk 3]
```diff
@@ -37,17 +36,11 @@ module.exports = async (
 
   // extract all the exports
   const nodeExports = extractExports(code);
+  const { frontmatter } = nodeExports;
```
Comment (ChristopherBiscardi) [inline packages/gatsby-mdx/on-create-node.js]: would need to make this change here too -- https://github.com/ChristopherBiscardi/gatsby-mdx/blob/master/packages/gatsby-mdx/utils/create-mdx-node.js

File: packages/gatsby-mdx/utils/extract-exports.js [hunk 4]
```diff
@@ -13,18 +14,16 @@ module.exports = code => {
...
+  // grab the frontmatter
+  const classicFrontmatter = exportedVariables._frontmatter || {};
+  const exportFrontmatter = exportedVariables.frontmatter || {};
+
+  // delete the frontmatter from the exports
+  delete exportedVariables._frontmatter;
+  delete exportedVariables.frontmatter;
+
+  // add the merged frontmatter to the exports
+  exportedVariables.frontmatter = merge(classicFrontmatter, exportFrontmatter);
```
Comment (ChristopherBiscardi) [inline packages/gatsby-mdx/utils/extract-exports.js]: Why are we importing lodash here? Can we use obj/rest spread instead?

**Comment by ChristopherBiscardi** [comment]
I'm going to push a couple syntax changes to get rid of lodash and then merge this

**Comment by KyleAMathews** [comment]
This looks great! I wonder if adding an API would make sense at some point for setting context on a page. Probably should wait for more use cases to emerge since what you're doing works well enough but something to think about.

**Comment by ChristopherBiscardi** [comment]
@KyleAMathews yeah, I think we'll wait for use cases to arise for that. I want to push people toward defining their own createPage() generally speaking (and we also need to update the page support in gatsby-mdx to use the new graphql approach first anyway)

**Comment by KyleAMathews** [comment]
Yeah, I was wondering if there'd be other cases where plugins would want to decorate pages with data. Seems like it could be useful.

**Comment by apuravchauhan** [comment]
@ChristopherBiscardi Can you please update the npm with this feature? Its still at 0.0.11

**Comment by ChristopherBiscardi** [comment]
@apuravchauhan I published this on stream today as 0.0.12. Current version is 0.0.13.

**Comment by apuravchauhan** [comment]
Thanks

**Pull request merged**